readme.rst
Vivus
========

django CMS Vivus integration


HISTORY
CHANGELOG
=========

0.0.2 (2020-02-02)
------------------

* Django 2 compatibility

0.0.1 (2018-02-06)
------------------

* Initial project setup


ripiu/cmsplugin_vivus/models.py
from django.db import models
from django.utils.translation import ugettext_lazy as _

from cms.models import CMSPlugin
from filer.fields.file import FilerFileField


class VivusPlugin(CMSPlugin):
    '''Drawing animation on SVG'''

    TYPE_DELAYED = 'delayed'
    TYPE_SYNC = 'sync'
    TYPE_ONE_BY_ONE = 'oneByOne'
    TYPE_SCRIPT = 'script'
    TYPE_SCENARIO = 'scenario'
    TYPE_SCENARIO_SYNC = 'scenario-sync'
    TYPE_CHOICES = (
        (TYPE_DELAYED, _('Delayed start')),
        (TYPE_SYNC, _('Synchronous')),
        (TYPE_ONE_BY_ONE, _('One by one')),
        # (TYPE_SCRIPT, TYPE_SCRIPT),
        # (TYPE_SCENARIO, TYPE_SCENARIO),
        # (TYPE_SCENARIO_SYNC, TYPE_SCENARIO_SYNC),
    )

    START_IN_VIEWPORT = 'inViewport'
    START_MANUAL = 'manual'
    START_AUTOSTART = 'autostart'
    START_CHOICES = (
        (START_IN_VIEWPORT, _('Start once the SVG is in the viewport')),
        (START_MANUAL, _('Start manually')),
        (START_AUTOSTART, _('Start right away')),
    )

    animation_type = models.CharField(
        _('animation type'), max_length=15, choices=TYPE_CHOICES,
        blank=False, default=TYPE_DELAYED,
        help_text=_('Defines what kind of animation will be used.')
    )

    svg_file = FilerFileField(
        blank=False, null=False, verbose_name=_('SVG file'),
        on_delete=models.CASCADE
    )

    start = models.CharField(
        _('animation trigger'), max_length=15, choices=START_CHOICES,
        blank=False, default=START_IN_VIEWPORT,
        help_text=_('Defines how to trigger the animation.')
    )

    duration = models.PositiveSmallIntegerField(
        _('duration'),
        blank=False, null=False, default=200,
        help_text=_('Animation duration, in frames.')
    )

    delay = models.PositiveSmallIntegerField(
        _('delay'),
        blank=False, null=False, default=0,
        help_text=_('Time between the drawing of first and last path, in '
                    'frames (only for delayed animations).')
    )

    # on_ready
    # Function called when the instance is ready to play.

    # path_timing_function
    # Timing animation function for each path element of the SVG.

    # anim_timing_function
    # Timing animation function for the complete SVG.

    dash_gap = models.PositiveSmallIntegerField(
        _('dash gap'),
        blank=False, null=False, default=2,
        help_text=_('Whitespace extra margin between dashes. Increase it in '
                    'case of glitches at the initial state of the animation.')
    )

    force_render = models.NullBooleanField(
        _('force render'), default=None,
        help_text=_('Force the browser to re-render all updated path items. '
                    'By default, the value is true on IE only.')
    )

    reverse_stack = models.BooleanField(
        _('reverse stack'), default=False,
        help_text=_("Reverse the order of execution. The default behaviour "
                    "is to render from the first 'path' in the SVG to the "
                    "last one. This option allow you to reverse the order.")
    )

    self_destroy = models.BooleanField(
        _('self destroy'), default=False,
        help_text=_('Removes all extra styling on the SVG, and leaves it as '
                    'original.')
    )

    def __str__(self):
        return str(self.svg_file)

    class Meta:
        verbose_name = _('Drawing animation on SVG')
        verbose_name_plural = _('Drawing animations on SVG')


ripiu/cmsplugin_vivus/cms_plugins.py
from django.utils.translation import ugettext_lazy as _

from cms.plugin_base import CMSPluginBase
from cms.plugin_pool import plugin_pool

from . import __version__
from .models import VivusPlugin


@plugin_pool.register_plugin
class VivusPluginPublisher(CMSPluginBase):
    model = VivusPlugin
    name = _('Vivus animation')
    module = 'Ri+'
    render_template = 'ripiu/cmsplugin_vivus/vivus.html'
    allow_children = True

    def render(self, context, instance, placeholder):
        import json
        context = super(VivusPluginPublisher, self).render(
            context, instance, placeholder
        )
        vivus_conf = {
            'animation_type': instance.animation_type,
            'start': instance.start,
            'duration': instance.duration,
            'delay': instance.delay,
            # 'onReady':
            # 'pathTimingFunction':
            # 'animTimingFunction':
            'dash_gap': instance.dash_gap,
            'reverse_stack': json.dumps(instance.reverse_stack),
            'self_destroy': json.dumps(instance.self_destroy),
        }
        if instance.force_render is not None:
            vivus_conf['force_render'] = json.dumps(instance.force_render)
        context.update({
            'instance': instance,
            'placeholder': placeholder,
            'version': __version__,
            'conf': vivus_conf,
        })
        return context


ripiu/cmsplugin_vivus/static/ripiu/cmsplugin_vivus/js/vivus.js
"use strict";

(function () {
    var VivusAnimation = (function () {
        /**
         * Helper function that retrieves a data-attribute value.
         *
         * @function getAttr
         * @param {HTMLElement} element single document node
         * @param {String} data data-attribute to retrieve
         * @return {String} value returns the value from the data-attribute
         */
        function getAttr(element, data) {
            var value = element.getAttribute("data-" + data);

            // true/false values need to be parsed from string to boolean
            // from the attributes data
            if (value === "true") {
                return true;
            } else if (value === "false") {
                return false;
            } else if (value === "null") {
                return null;
            }
            return value;
        }

        /**
         * Configures and starts Vivus.
         *
         * @class VivusAnimation
         * @constructor
         * @param {HTMLElement} container single document node
         */
        function VivusAnimationConstructor(container) {
            this.container = container;
            this.settings = {
                type: getAttr(container, "animation-type"),
                start: getAttr(container, "start"),
                duration: parseInt(getAttr(container, "duration")),
                delay: parseInt(getAttr(container, "delay")),
                dashGap: parseInt(getAttr(container, "dash-gap")),
                forceRender: getAttr(container, "force-render"),
                reverseStack: getAttr(container, "reverse-stack"),
                selfDestroy: getAttr(container, "self-destroy")
            };
            console.log(this.settings);
            new Vivus(this.container, this.settings);
        }
        return VivusAnimationConstructor;
    })();

    window.addEventListener("load", function () {
        var elements = document.getElementsByClassName("ripiu-vivus");
        elements = [].slice.call(elements);
        elements.forEach(function (element) {
            new VivusAnimation(element);
            // new Vivus(element, {duration: 200});
        })
    })
})();


ripiu/cmsplugin_vivus/migrations/0001_initial.py
# -*- coding: utf-8 -*-
# Generated by Django 1.10.8 on 2018-02-06 10:50
from __future__ import unicode_literals

from django.db import migrations, models
import django.db.models.deletion
import filer.fields.file


class Migration(migrations.Migration):

    initial = True

    dependencies = [
        ('cms', '0016_auto_20160608_1535'),
        ('filer', '0007_auto_20161016_1055'),
    ]

    operations = [
        migrations.CreateModel(
            name='VivusPlugin',
            fields=[
                ('cmsplugin_ptr', models.OneToOneField(auto_created=True, on_delete=django.db.models.deletion.CASCADE, parent_link=True, primary_key=True, related_name='cmsplugin_vivus_vivusplugin', serialize=False, to='cms.CMSPlugin')),
                ('animation_type', models.CharField(choices=[('delayed', 'Delayed start'), ('sync', 'Synchronous'), ('oneByOne', 'One by one')], default='delayed', help_text='Defines what kind of animation will be used.', max_length=15, verbose_name='animation type')),
                ('start', models.CharField(choices=[('inViewport', 'Start once the SVG is in the viewport'), ('manual', 'Start manually'), ('autostart', 'Start right away')], default='inViewport', help_text='Defines how to trigger the animation.', max_length=15, verbose_name='animation trigger')),
                ('duration', models.PositiveSmallIntegerField(default=200, help_text='Animation duration, in frames.', verbose_name='duration')),
                ('delay', models.PositiveSmallIntegerField(default=0, help_text='Time between the drawing of first and last path, in frames (only for delayed animations).', verbose_name='delay')),
                ('dash_gap', models.PositiveSmallIntegerField(default=2, help_text='Whitespace extra margin between dashes. Increase it in case of glitches at the initial state of the animation.', verbose_name='dash gap')),
                ('force_render', models.NullBooleanField(default=None, help_text='Force the browser to re-render all updated path items. By default, the value is true on IE only.', verbose_name='force render')),
                ('reverse_stack', models.BooleanField(default=False, help_text="Reverse the order of execution. The default behaviour is to render from the first 'path' in the SVG to the last one. This option allow you to reverse the order.", verbose_name='reverse stack')),
                ('self_destroy', models.BooleanField(default=False, help_text='Removes all extra styling on the SVG, and leaves it as original.', verbose_name='self destroy')),
                ('svg_file', filer.fields.file.FilerFileField(on_delete=django.db.models.deletion.CASCADE, to='filer.File', verbose_name='SVG file')),
            ],
            options={
                'verbose_name': 'Drawing animation on SVG',
                'verbose_name_plural': 'Drawing animations on SVG',
            },
            bases=('cms.cmsplugin',),
        ),
    ]
